src/cli/expand-patterns.js

```diff
@@ -0,0 +1,221 @@
+"use strict";
+
+const path = require("path");
+const fs = require("fs");
+const globby = require("globby");
+const flat = require("lodash/flatten");
+
+/** @typedef {import('./util').Context} Context */
+
+/**
+ * @param {Context} context
+ */
+function* expandPatterns(context) {
+  const cwd = process.cwd();
+  const seen = Object.create(null);
+  let noResults = true;
+
+  for (const pathOrError of expandPatternsInternal(context)) {
+    noResults = false;
+    if (typeof pathOrError !== "string") {
+      yield pathOrError;
+      continue;
+    }
+
+    const relativePath = path.relative(cwd, pathOrError);
+
+    // filter out duplicates
+    if (relativePath in seen) {
+      continue;
+    }
+
+    seen[relativePath] = true;
+    yield relativePath;
+  }
+
+  if (noResults) {
+    // If there was no files and no other errors, let's yield a general error.
+    yield {
+      error: `No matching files. Patterns: ${context.filePatterns.join(" ")}`
+    };
+  }
+}
+
+const isWindows = path.sep === "\\";
+
+// TODO: use `fast-glob` directly, without `globby`
+const baseGlobbyOptions = {
+  dot: true,
+  expandDirectories: false,
```

Comment:
Do we  support expand dirs in pattern? Like  `prettier **/foo` should work on any files in dir  named foo.  `expandDirectories` should work

Reply:
ESLint does this. Looks like we should too. But I doubt we can use globby's `expandDirectories` for that. Our expansion process is too custom, especially if we add parser inference to it.

Reply:
Sorry, I was wrong. ESLint actually **doesn't** support this. I forgot to put quotes around the glob, so it was resolved by Bash. So do we want to add this?

Reply:
I am almost sure that we will receive a request for this in the near future with real-life examples, so will be great to support these patterns, if it does not require much effort, if it requires, we can postpone it

Reply:
This use case doesn't look very useful to me. What tools support this?

Reply:
hm, eslint doesn't support, stylelint doesn't support, maybe @fisker known tools what supported that

Reply:
If eslint and stylelint doesn't support, I'm ok we don't support this